File: dso3000.py
import usb

class DSO3000:
	'''Class to interface over USB with a Agilent DSO3000A-series (and probably Rigol DSO5000) oscilloscope.
	
	Example:
	  scope = DSO3000()    # Connect to scope
	  print scope.command(':CHAN1:COUPL?')   # Print selected channel
	  print scope.readScreen(1)              # Read waveform in volts
	'''
	
	def __init__(self, flush=True):
		'''Finds and opens the scope's USB device.
		If flush==True, any pending data is discarded.'''
		
		self.device = None
		
		# Time in milliseconds to wait for a USB control transaction to finish.
		# *RST takes about 1500ms, so the timeout should be longer than that.
		self.timeout = 10000
		
		for bus in usb.busses():
			for dev in bus.devices:
				if dev.idVendor == 0x0400 and dev.idProduct == 0xc55d:
					self.device = dev.open()
					self.device.setConfiguration(1)
					self.device.claimInterface(0)
					break
		if not self.device:
			raise IOError('No USB oscilloscope found')
		if flush:
			self.read()
	
	def writeChar(self, ch):
		'''Writes a single character to the scope'''
		assert self.device.controlMsg(0xc0, 1, 0, ord(ch), 0, self.timeout) == ()
	
	def write(self, s):
		'''Writes a string and an end-of-line to the scope'''
		for ch in s:
			self.writeChar(ch)
		self.writeChar('\r')

	def getResponseLength(self):
		'''Returns the number of characters waiting to be read.
		If more than 255 characters are available, returns 255.'''
		return self.device.controlMsg(0xc0, 0, 1, 0, 0, self.timeout)[0]
	
	def read(self):
		'''Reads as much data as is currently available'''
		text = ''
		while True:
			n = self.getResponseLength()
			if n == 0:
				break
			data = self.device.controlMsg(0xc0, 0, n, 1, 0, self.timeout)
			text += ''.join(chr(x) for x in data).rstrip('\r\n')
		
		return text
	
	def command(self, s):
		'''Writes a command string and returns the response (if any)'''
		
		# Write the command
		self.write(s)
		
		# Read the response
		text = self.read()
		if not text:
			# No response text
			return None

		# When requesting waveform data, the firmware sends a
		# bunch of junk in memory that we have to discard.
		# Keep everything up to the first newline.
		return text.splitlines()[0]
	
	def readData(self, command):
		'''Reads waveform data with the given command and returns the waveform in volts.'''
		# Send the command
		text = self.command(command)
		
		# Parse the result into ADC readings
		data = [int(x, 16) for x in text.split()]
		
		# Get channel configuration needed to convert to volts
		yinc = float(self.command(':WAV:YINC?'))
		yor = float(self.command(':WAV:YOR?'))
		
		# Convert ADC readings to volts.
		# Agilent's Programmer's Reference appears to have this formula right,
		# but setting a channel to GND produces all 126's (one count below zero volts).
		data = [(125.0 - x) * yinc - yor for x in data]
		return data
	
	def readMemory(self, channel):
		'''Reads waveform memory for the given channel (1 or 2)'''
		self.command(':WAV:SOUR CHANNEL%d' % channel)
		return self.readData(':WAV:MEM?')

	def readScreen(self, channel):
		'''Reads the waveform on the screen for the given channel (1 or 2)'''
		self.command(':WAV:SOUR CHANNEL%d' % channel)
		return self.readData(':WAV:DATA?')
	
	def screenshot(self):
		self.write(':HARD_COPY')
		assert self.device.controlMsg(0xc0, 8, 0, 0, 0x50, self.timeout) == ()
		# wIndex:wValue is the length of data returned
		assert self.device.controlMsg(0xc0, 9, 0, 0x2c00, 1, self.timeout) == ()
		assert self.device.controlMsg(0xc0, 7, 0, 0, 0x50, self.timeout) == ()
		data = self.device.bulkRead(1, 76800, self.timeout)
		assert self.getResponseLength() == 0
		return data

# Colors:
# 50 background yellow
# 30 Remote red/pink
# c0 green
# a8 white
# e0 yellow
# f0 blue
# inverted changes the meaning of e0...

File: dso3000.cpp
#include "dso3000.hpp"

#include <libusb.h>
#include <stdexcept>

using namespace std;

DSO3000::DSO3000()
{
	_usb = 0;
	_timeout = 10000;

	if (libusb_init(&_usb))
	{
		throw runtime_error("Failed to initialize libusb");
	}

	_device = libusb_open_device_with_vid_pid(_usb, 0x0400, 0xc55d);
	if (!_device)
	{
		throw runtime_error("Failed to open USB device");
	}

	if (libusb_set_configuration(_device, 1))
	{
		throw runtime_error("Can't set configuration");
	}

	if (libusb_claim_interface(_device, 0))
	{
		throw runtime_error("Can't claim interface");
	}

	flush();
}

DSO3000::~DSO3000()
{
	if (_device)
	{
		libusb_close(_device);
	}

	libusb_exit(_usb);
}

int DSO3000::flush()
{
	char buf[256];
	int len, total = 0;

	do
	{
		len = read(buf, sizeof(buf));
		total += len;
	} while (len);

	return total;
}

void DSO3000::writeChar(char ch)
{
	libusb_control_transfer(_device, 0xc0, 1, ch, 0, 0, 0, _timeout);
}

void DSO3000::write(const char *str)
{
	for (int i = 0; str[i]; i++)
	{
		writeChar(str[i]);
	}
	writeChar('\r');
}

int DSO3000::getResponseLength()
{
	unsigned char len;

	libusb_control_transfer(_device, 0xc0, 0, 0, 0, &len, 1, _timeout);

	return len;
}

int DSO3000::read(char *buf, int maxlen)
{
	int i, len;

	len = getResponseLength();
	if (!len)
	{
		if (maxlen)
			buf[0] = 0;

		return 0;
	}

	if (len > (maxlen - 1))
	{
		len = maxlen - 1;
	}

	len = libusb_control_transfer(_device, 0xc0, 0, 1, 0, (unsigned char *)buf, len, _timeout);

	// Truncate the response at the first '\n' because the scope firmware sometimes
	// sends extra junk.
	for (i = 0; i < len; i++)
	{
		if (buf[i] == '\n')
		{
			len = i;
			break;
		}
	}
	buf[len] = 0;

	return len;
}

bool DSO3000::screenshot(unsigned char* buf)
{
	write(":HARD_COPY");
	
	if (libusb_control_transfer(_device, 0xc0, 8, 0, 0x50, 0, 0, _timeout))
	{
		return false;
	}
	
	// wIndex:wValue is the length of data returned
	if (libusb_control_transfer(_device, 0xc0, 9, 0x2c00, 1, 0, 0, _timeout))
	{
		return false;
	}
	
	if (libusb_control_transfer(_device, 0xc0, 7, 0, 0x50, 0, 0, _timeout))
	{
		return false;
	}
	
	int len = 0;
	if (libusb_bulk_transfer(_device, 0x81, buf, 76800, &len, _timeout))
	{
		return false;
	}
	
	getResponseLength();
	
	return len == 76800;
}
